exchange/index.go

```diff
@@ -453,6 +490,165 @@ func (idx *Index) evict(size uint64) uint64 {
+		return idx.bstore.DeleteBlock(c)
+	})
+	if err != nil {
+		log.Error().Err(err).Msg("failed to run garbage collector")
+	}
+
+	idx.evictionSet = cid.NewSet()
+	idx.lastEvictionTime = time.Now()
+}
+
+func (idx *Index) gcLoop() {
+	go func() {
+		ticker := time.NewTicker(idx.gcLoopDuration)
+
+		for {
+			select {
+			case now := <-ticker.C:
+				// if an eviction happened within the last gcLoopDuration, no need to do one now
+				if idx.lastEvictionTime.Add(idx.gcLoopDuration).After(now) {
+					continue
+				}
+
+				idx.GC()
+			}
+		}
+	}()
+}
+
+func (idx *Index) cleanBlockStore() error {
+	idx.emu.Lock()
+	defer idx.emu.Unlock()
+
+	cidSet := cid.NewSet()
+
+	err := idx.root.ForEach(context.Background(), func(k string, val *cbg.Deferred) error {
+		ref := new(DataRef)
+		err := ref.UnmarshalCBOR(bytes.NewReader(val.Raw))
+		if err != nil {
+			return err
+		}
+
+		err = idx.browseDAG(ref.PayloadCID, func(blk blocks.Block) error {
+			cidSet.Add(blk.Cid())
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	kc, err := idx.Bstore().AllKeysChan(context.Background())
+	for k := range kc {
+		if cidSet.Has(k) {
+			continue
+		}
+
+		err = idx.Bstore().DeleteBlock(k)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (idx *Index) browseDAG(root cid.Cid, f func(blocks.Block) error) error {
+	link := cidlink.Link{Cid: root}
+	chooser := dagpb.AddDagPBSupportToChooser(func(ipld.Link, ipld.LinkContext) (ipld.NodePrototype, error) {
+		return basicnode.Prototype.Any, nil
+	})
+
+	// The root node could be a raw node so we need to select the builder accordingly
+	nodeType, err := chooser(link, ipld.LinkContext{})
+	if err != nil {
+		return err
+	}
+	builder := nodeType.NewBuilder()
+
+	// We make a custom loader to intercept when each block is read during the traversal
+	makeLoader := func(bs blockstore.GCBlockstore) ipld.Loader {
+		return func(lnk ipld.Link, lnkCtx ipld.LinkContext) (io.Reader, error) {
+			c, ok := lnk.(cidlink.Link)
+			if !ok {
+				return nil, fmt.Errorf("incorrect Link Type")
+			}
+
+			block, err := bs.Get(c.Cid)
+			if err != nil {
+				return nil, err
+			}
+
+			reader := bytes.NewReader(block.RawData())
+			err = f(block)
+			if err != nil {
+				return nil, err
+			}
+
+			return reader, nil
+		}
+	}
+
+	// Load the root node
+	err = link.Load(context.TODO(), ipld.LinkContext{}, builder, makeLoader(idx.bstore))
+	if err != nil {
+		return fmt.Errorf("unable to load link: %v", err)
+	}
+	nd := builder.Build()
+
+	s, err := selector.ParseSelector(sel.All())
+	if err != nil {
+		return err
+	}
+
+	// Traverse any links from the root node
+	err = traversal.Progress{
+		Cfg: &traversal.Config{
+			LinkLoader:                     makeLoader(idx.bstore),
+			LinkTargetNodePrototypeChooser: chooser,
+		},
+	}.WalkMatching(nd, s, func(prog traversal.Progress, n ipld.Node) error {
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
I tried to make function to iterate over a dag, but I think something is wrong, it seems that on my test it can't decode it properly : ```No automatic decoding for this type, node builder must provide fast path```

Reply:
@tchardin do you know what the error means? I don't see what's wrong on my walkDAG function, here is the tests that displays the error :
https://github.com/myelnet/pop/blob/ag/feat-garbage-collection-strategy/exchange/index_test.go#L581

Reply:
```
{"level":"error","error":"unable to load link: unmarshal failed. proto: PBNode: illegal tag 0 (wire type 1)","time":"2021-06-30T10:16:30+02:00","message":"failed to tag ref QmTTA2daxGqo5denp6SwLzzkLJm3fuisYEi9CoWsuHpzfb for eviction"}
```

Reply:
hum I think it's because of the block generation function i'm using `b1 := blockGen.Next()` because it's only blocks made of increasing numbers, it doesn't like the 0 encode to a block

Reply:
It was because I was using a Block instead of a Node for the tests